sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -409,16 +416,34 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     }
 
     if (DDLUtils.isDatasourceTable(withStatsProps)) {
-      val oldDef = client.getTable(db, withStatsProps.identifier.table)
-      // Sets the `schema`, `partitionColumnNames` and `bucketSpec` from the old table definition,
-      // to retain the spark specific format if it is. Also add old data source properties to table
-      // properties, to retain the data source table format.
-      val oldDataSourceProps = oldDef.properties.filter(_._1.startsWith(DATASOURCE_PREFIX))
+      val oldTableDef = client.getTable(db, withStatsProps.identifier.table)
+
+      val oldLocation = if (tableDefinition.tableType == EXTERNAL) {
+        oldTableDef.properties.get(DATASOURCE_LOCATION)
+      } else {
+        tableDefinition.storage.locationUri
+      }
+      // Only update the `locationUri` field if the location is really changed.
+      val newLocation = if (oldLocation == tableDefinition.storage.locationUri) {
```

Comment:
This is to keep the previous behaviour.

Previously, if we alter table fields (exclude location), we will keep the `locationUri` field and `path` option unchanged. If we alter table location, we will update both the `locationUri` field and `path` option, although it's dangerous to set `locationUri` as the path may be a file path.

Now, if we alter table fields (exclude location), we will keep the `locationUri` field unchanged, although the `locationUri` field may be different from the one in old raw table(the `oldTableDef` get from hive client directly) due to the hacks in `saveTableIntoHive`. If we alter table location, we will update the `locationUri` field even it's file path, which is the same behaviour as before.